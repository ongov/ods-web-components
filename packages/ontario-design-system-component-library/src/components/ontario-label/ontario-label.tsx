import { Component, h, Element, Prop, Watch } from '@stencil/core';

/**
 * Ontario Label component properties
 */
export interface OntarioLabelProperties {
	/**
	 * The text to display as label.
	 */
	caption?: string;

	/**
	 * The form control with which the caption is associated.
	 */
	for: string;

	/**
	 * Defines whether the input field is required. If required, the value passed should be 'required'.
	 */
	required?: boolean;

	/**
	 * The type of label to render.
	 */
	type?: 'default' | 'large' | 'heading';
}

/**
 * Ontario Design System label web component
 */
@Component({
	tag: 'ontario-label',
	styleUrl: 'ontario-label.scss',
	shadow: true,
})
export class OntarioLabel implements OntarioLabelProperties {
	@Element() host: HTMLElement;

	/**
	 * The text to display as label. This will override the text provided through the Element Content.
	 *
	 * @example
	 * <ontario-label caption="Comments" for"comments">Feedback</ontario-label>
	 *
	 * The resulting label will show `"Comments"`.
	 */
	@Prop({ mutable: true }) caption?: string;

	/**
	 * The form control with which the caption is associated.
	 */
	@Prop() for: string;

	/**
	 * The type of label to render.
	 */
	@Prop() type: 'default' | 'large' | 'heading' = 'default';

	/**
	 * Defines whether the input field is required. If required, the value passed should be 'required'.
	 */
	@Prop() required?: boolean = false;

	@Watch('caption')
	validateCaption(newValue: string) {
		const isCaptionBlank = typeof newValue !== 'string' || newValue === '';
		const isElementContentBlank = typeof this.host.textContent !== 'string' || this.host.textContent === '';
		if (isCaptionBlank && isElementContentBlank) {
			throw new Error('Label cannot be empty. Please set the label through its caption property or by inserting text between the <ontario-label> tags.');
		}
	}

	/**
	 * Set `caption` using internal component logic
	 */
	componentWillLoad() {
		this.caption = this.caption ?? this.host.textContent ?? '';
	}

	/**
	 *
	 * @returns the classes of the label based on the label's `type`
	 */
	private getClass() {
		return this.type.toLowerCase() === 'default' ? `ontario-label` : `ontario-label ontario-label--${this.type.toLowerCase()}`;
	}

	/**
	 *
	 * @returns the flag text based on whether the input is required or optional
	 */
	private getRequiredFlagText() {
		return this.required ? '(required)' : '(optional)';
	}

	/**
	 *
	 * @returns the label element to be rendered by the browser
	 */
	private getLabelElement() {
		const labelContent = (
			<label htmlFor={this.for} class={this.getClass()}>
				{this.caption}
				<span class="ontario-label__flag">{this.getRequiredFlagText()}</span>
			</label>
		);
		return this.type.toLowerCase() === 'heading' ? <h1>{labelContent}</h1> : labelContent;
	}

	render() {
		return this.getLabelElement();
	}
}
