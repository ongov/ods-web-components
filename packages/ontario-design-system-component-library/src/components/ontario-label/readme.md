# ontario-label



<!-- Auto Generated Below -->


## Properties

| Property   | Attribute  | Description                                                                                      | Type                                | Default     |
| ---------- | ---------- | ------------------------------------------------------------------------------------------------ | ----------------------------------- | ----------- |
| `caption`  | `caption`  | The text to display as label. This will override the text provided through the Element Content.  | `string`                            | `undefined` |
| `for`      | `for`      | The form control with which the caption is associated.                                           | `string`                            | `undefined` |
| `required` | `required` | Defines whether the input field is required. If required, the value passed should be 'required'. | `boolean \| undefined`              | `false`     |
| `type`     | `type`     | The type of label to render.                                                                     | `"default" \| "heading" \| "large"` | `'default'` |


----------------------------------------------

*Built with [StencilJS](https://stenciljs.com/)*
