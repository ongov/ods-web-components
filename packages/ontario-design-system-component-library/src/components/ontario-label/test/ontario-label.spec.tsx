import { newSpecPage } from '@stencil/core/testing';
import { OntarioLabel } from '../ontario-label';

describe('ontario-label snapshot', () => {
	it('should render the expected html of ontario-label', async () => {
		const page = await newSpecPage({
			components: [OntarioLabel],
			html: `<ontario-label>Default Label</ontario-label>`,
		});
		expect(page.root).toMatchSnapshot();
	});
});

describe('ontario-label', () => {
	it('should render the default label element', async () => {
		const page = await newSpecPage({
			components: [OntarioLabel],
			html: `<ontario-label>Default Label</ontario-label>`,
		});
		expect(page.root).toEqualHtml(`
      		<ontario-label>
        		<mock:shadow-root>
          			<label class="ontario-label">Default Label
            			<span class="ontario-label__flag">(optional)</span>
          			</label>
        		</mock:shadow-root>
        		Default Label
      		</ontario-label>
    	`);
	});

	it('should render the large label element', async () => {
		const page = await newSpecPage({
			components: [OntarioLabel],
			html: `<ontario-label type="large">Large Label</ontario-label>`,
		});
		expect(page.root).toEqualHtml(`
      		<ontario-label type="large">
        		<mock:shadow-root>
          			<label class="ontario-label ontario-label--large">Large Label
            			<span class="ontario-label__flag">(optional)</span>
          			</label>
        		</mock:shadow-root>
        		Large Label
      		</ontario-label>
    	`);
	});

	it('should render the heading label element', async () => {
		const page = await newSpecPage({
			components: [OntarioLabel],
			html: `<ontario-label type="heading">Heading Label</ontario-label>`,
		});
		expect(page.root).toEqualHtml(`
      		<ontario-label type="heading">
        		<mock:shadow-root>
          			<h1>
            			<label class="ontario-label ontario-label--heading">Heading Label
              				<span class="ontario-label__flag">(optional)</span>
            			</label>
          			</h1>
        		</mock:shadow-root>
        		Heading Label
      		</ontario-label>
    	`);
	});

	it('should render a default required label element', async () => {
		const page = await newSpecPage({
			components: [OntarioLabel],
			html: `<ontario-label required>Default Label</ontario-label>`,
		});
		expect(page.root).toEqualHtml(`
      		<ontario-label required>
        		<mock:shadow-root>
          			<label class="ontario-label">Default Label
            			<span class="ontario-label__flag">(required)</span>
          			</label>
        		</mock:shadow-root>
        		Default Label
      		</ontario-label>
    	`);
	});

	it('should render a default required label element', async () => {
		const page = await newSpecPage({
			components: [OntarioLabel],
			html: `<ontario-label required>Default Label</ontario-label>`,
		});
		expect(page.root).toEqualHtml(`
      <ontario-label required>
        <mock:shadow-root>
          <label class="ontario-label">Default Label
            <span class="ontario-label__flag">(required)</span>
          </label>
        </mock:shadow-root>
        Default Label
      </ontario-label>
    `);
	});

	it('should render a default label element with caption overriding the element content', async () => {
		const page = await newSpecPage({
			components: [OntarioLabel],
			html: `<ontario-label caption="Label Caption">Default Label</ontario-label>`,
		});
		expect(page.root).toEqualHtml(`
    		<ontario-label caption="Label Caption">
        		<mock:shadow-root>
          			<label class="ontario-label">Label Caption
            			<span class="ontario-label__flag">(optional)</span>
          			</label>
        		</mock:shadow-root>
        		Default Label
      		</ontario-label>
    	`);
	});

	it('should render a default label element setting the htmlFor attribute', async () => {
		const page = await newSpecPage({
			components: [OntarioLabel],
			html: `<ontario-label for="comments">Comments</ontario-label>`,
		});
		expect(page.root).toEqualHtml(`
      		<ontario-label for="comments">
        		<mock:shadow-root>
          			<label class="ontario-label" htmlFor="comments">Comments
            			<span class="ontario-label__flag">(optional)</span>
          			</label>
        		</mock:shadow-root>
        		Comments
      		</ontario-label>
    	`);
	});

	it('should throw an error when label is not supplied', async () => {
		try {
			await newSpecPage({
				components: [OntarioLabel],
				html: `<ontario-label></ontario-label>`,
			});
		} catch (validationError) {
			expect(validationError).toEqual(new Error('Label cannot be empty. Please set the label through its caption property or by inserting text between the <ontario-label> tags.'));
		}
	});
});
