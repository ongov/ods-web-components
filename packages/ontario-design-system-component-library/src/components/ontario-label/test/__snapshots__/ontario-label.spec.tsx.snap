// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ontario-label snapshot should render the expected html of ontario-label 1`] = `
<ontario-label>
  <mock:shadow-root>
    <label class="ontario-label">
      Default Label
      <span class="ontario-label__flag">
        (optional)
      </span>
    </label>
  </mock:shadow-root>
  Default Label
</ontario-label>
`;
